design/ozone.md

```diff
@@ -0,0 +1,246 @@
+ 2. Additional attributes can trigger additional transformations
+ 3. Finally we can provide an option to define any custom transformations
+
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: rook-ozone
+---
+apiVersion: ozone.rook.io/v1alpha1
+kind: ObjectStore
+metadata:
+  name: my-store
+  namespace: rook-ozone
+spec:
+  components: #high lebel options to turn off/on component groups
+     csi: true
+     recon: true
+     s3g: true
+  features: # these are not components but confiuration groups
+     prometheus: true
+     tracing: true
+  ozoneVersion: #see ceph versioning doc for mor details
+    image: apache/ozone:0.4.1
+    allowUnsupported: false
+  upgradePolicy: #see ceph versioning doc for mor details (Won't be implmented in Phase 1)
+    ozoneVersion:
+      image: apache/ozone:0.5.0
+      allowUnsupported: false
+  storage: #more storage option should be added to here
+     type: ephemeral/auto/manual # ephemeral: use emptyDir, auto: using the existing rook method to identify local disk, manual: should be specified
+     persistence:
+        metadata: #specify persistent volumes for metadata disk
+        storage: #specify persistent volumes for storage
+  annotations: #custom annotations for all the created k8s resources
+     key: value
+  customization: #deep water: any additional transformations to modify k8s resources. Only for advanced users.
+    - type: Add
+      path: ["metadata", "labels"]
+      value:
+         generated-with: rook
+    
+```
+
+## Operator capabilities
+
+The main benefit to use an operator is to provide additional functionalities based on the current state of the runtime environment. The power of the operator based on the runtime information which is not available for any other installer tool.
+
+Some option which can be implemented as part of the Operator:
+
+ * If only one node is available (eg. Minikube) in the cluster, use `StatefulSet` with `replicas: 3` to make it possible to test full replicatoin. 
+ * Use emptyDir instead of Persistent Volumes in case of the dev cluster (_opt-in_ only, it should be requtested by a CRD attribute)
+ * Connection kubernetes decommissioning and Ozone decommissioning
+ * Configure local persistence based on the available disks
+ * Support storage based auto-scaling (if datanodes are `StatefulSet`)
+ * Expose S3 endpoint based on the avaialble ingress, LB
+ * Volume support: Ozone has a hierarchy of volumes/buckets/keys. Bucket and keys are the same what we have in any S3 compatible store, volumes can contain generic replication/security rules for all the assigned buckets. We can support Volume CRD to make it easier the managemenet of the volumes
+ * Environment specific configuration (host names) can be adjusted based on the current topology (host names, namespace)
+ * Based on the current size of the cluster better default settings can be provided (memory settings, block size, etc.)
```

Comment:
Resource limits and requests are good to specify in the CR so the user can have control.

Reply:
I removed, but some background about my thinking:

I don't think that an average user specifies the resource limits. The big question here is the default. Can we provide smart defaults based on the available resources?

It may not be required, as `resources.limits` can be optional and `resources.requests` can have meaningful default values.